leader/undirected_ring/async_itai_rodeh/async_itai_rodeh.go

```diff
@@ -0,0 +1,242 @@
+package async_itai_rodeh
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"log"
+	"math"
+	"math/rand"
+
+	"github.com/krzysztof-turowski/distributed-framework/lib"
+)
+
+const (
+	LEFT  = 0
+	RIGHT = 1
+)
+
+type state struct {
+	Counter          int
+	Is_active        bool
+	Is_leader        bool
+	Is_leader_chosen bool
+	Bit              bool
+}
+
+type messageData struct {
+	Msg_type  uint8
+	Bit       bool
+	Hop_count uint64
+}
+
+//Ring start and answer check
+
+func Run(nodes []lib.Node, runner lib.Runner) (int, int) {
+	for _, node := range nodes {
+		log.Println("Running node", node.GetIndex())
+		go run(node)
+	}
+	runner.Run(true)
+
+	leaders := 0
+
+	for _, node := range nodes {
+		if getState(node).Is_leader {
+			leaders += 1
+		}
+	}
+	if leaders == 0 {
+		panic("No leader elected")
+	}
+	if leaders > 1 {
+		panic("More than one leader elected")
+	}
+
+	log.Println("---OK--- exactly one leader elected")
+
+	return runner.GetStats()
+}
+
+// Main function run in each goroutine
+
+func run(node lib.Node) {
+	node.StartProcessing()
+	initialize(node)
+	if process(node) {
+		log.Println("Node", node.GetIndex(), "is elected a leader")
+	}
+	node.FinishProcessing(true)
+}
+
+// Main algorithm loop
+
+func process(node lib.Node) bool {
+	for {
```

Comment:
Please refactor it so that one invocation of process = one ReceiveAnyMessage() call (just put the for loop in run())

Reply:
Done